Allow importing jars that don't have Smithy files

Sometimes, you just want to resolve an artifact, and pass its jar (and its dependencies' jars) to a model assembler. At the moment, this fails if any of these jars doesn't have a Smithy manifest file.

Could this constraint become more relaxed? For context, something similar has already happened with importing JSON files in #1846 - the non-smithy ones are ignored with just a log message. Can we do the same here?

Yes, as long as a jar doesn't have a Smithy manifest, trying to import it throws an exception:

```scala
//> using dep software.amazon.smithy:smithy-model:1.69.0
import java.nio.file.Paths
import software.amazon.smithy.model.Model

@main def demo = {

  println(
    Model
      .assembler()
      .addImport(Paths.get("demo.jar").toUri().toURL())
      .assemble()
      .unwrap()
  )
}
```

(where `demo.jar` is a dummy jar without a manifest)

```
Exception in thread "main" software.amazon.smithy.model.loader.ModelManifestException: Error parsing Smithy model manifest from jar:file:///private/var/folders/f_/cbwjhbrs2jjg59qn23y6nhvm0000gn/T/tmp.Udq2XCzpOZ/demo.jar!/META-INF/smithy/manifest
        at software.amazon.smithy.model.loader.ModelDiscovery.findModels(ModelDiscovery.java:151)
        at software.amazon.smithy.model.loader.ModelLoader.loadJar(ModelLoader.java:121)
        at software.amazon.smithy.model.loader.ModelLoader.load(ModelLoader.java:66)
        at software.amazon.smithy.model.loader.ModelAssembler.assemble(ModelAssembler.java:547)
        at main$package$.demo(main.scala:10)
        at demo.main(main.scala:5)
Caused by: java.io.FileNotFoundException: JAR entry META-INF/smithy/manifest not found in /private/var/folders/f_/cbwjhbrs2jjg59qn23y6nhvm0000gn/T/tmp.Udq2XCzpOZ/demo.jar
        at java.base/sun.net.www.protocol.jar.JarURLConnection.connect(JarURLConnection.java:131)
        at java.base/sun.net.www.protocol.jar.JarURLConnection.getInputStream(JarURLConnection.java:160)
        at software.amazon.smithy.model.loader.ModelDiscovery.parseManifest(ModelDiscovery.java:220)
        at software.amazon.smithy.model.loader.ModelDiscovery.findModels(ModelDiscovery.java:145)
```

It would be preferable to just ignore such imports, this is especially useful if you have a list of jars as a result of dependency resolution. It might even be opt-in with `putProperty`.

--- a/.changes/next-release/feature-6dc2e6dcfbad136a8cffecc102b0f02214cd9e46.json
+++ b/.changes/next-release/feature-6dc2e6dcfbad136a8cffecc102b0f02214cd9e46.json
@@ -1,5 +1,7 @@
 {
   "type": "feature",
   "description": "Ignore JARs without a Smithy manifest to allow Smithy model packages to have non-Smithy dependencies",
-  "pull_requests": []
+  "pull_requests": [
+    "[#3055](https://github.com/smithy-lang/smithy/pull/3055)"
+  ]
 }

--- a/.changes/next-release/feature-6dc2e6dcfbad136a8cffecc102b0f02214cd9e46.json
+++ b/.changes/next-release/feature-6dc2e6dcfbad136a8cffecc102b0f02214cd9e46.json
@@ -1,0 +1,5 @@
+{
+  "type": "feature",
+  "description": "Ignore JARs without a Smithy manifest to allow Smithy model packages to have non-Smithy dependencies",
+  "pull_requests": []
+}

--- a/smithy-model/src/main/java/software/amazon/smithy/model/loader/ModelLoader.java
+++ b/smithy-model/src/main/java/software/amazon/smithy/model/loader/ModelLoader.java
@@ -4,10 +4,12 @@
  */
 package software.amazon.smithy.model.loader;
 
+import java.io.FileNotFoundException;
 import java.io.IOException;
 import java.io.InputStream;
 import java.net.URL;
 import java.net.URLConnection;
+import java.util.List;
 import java.util.Map;
 import java.util.function.Consumer;
 import java.util.function.Function;
@@ -118,7 +120,18 @@ final class ModelLoader {
         URL manifestUrl = ModelDiscovery.createSmithyJarManifestUrl(filename);
         LOGGER.fine(() -> "Loading Smithy model imports from JAR: " + manifestUrl);
 
-        for (URL model : ModelDiscovery.findModels(manifestUrl)) {
+        List<URL> models;
+        try {
+            models = ModelDiscovery.findModels(manifestUrl);
+        } catch (ModelManifestException e) {
+            if (e.getCause() instanceof FileNotFoundException) {
+                LOGGER.info(() -> "Ignoring JAR without Smithy manifest: " + filename);
+                return;
+            }
+            throw e;
+        }
+
+        for (URL model : models) {
             try {
                 URLConnection connection = model.openConnection();
 

--- a/smithy-model/src/test/java/software/amazon/smithy/model/loader/ModelAssemblerTest.java
+++ b/smithy-model/src/test/java/software/amazon/smithy/model/loader/ModelAssemblerTest.java
@@ -1348,6 +1348,16 @@ public class ModelAssemblerTest {
     }
 
     @Test
+    public void ignoresJarWithoutSmithyManifest() throws URISyntaxException, IOException {
+        Path jar = JarUtils.createJarFromDir(
+                Paths.get(getClass().getResource("assembler-no-manifest-jar").toURI()));
+
+        ValidatedResult<Model> result = Model.assembler().addImport(jar).assemble();
+
+        assertFalse(result.isBroken());
+    }
+
+    @Test
     public void doesNotThrowOnInvalidSuppression() {
         ObjectNode node = Node.objectNode()
                 .withMember("smithy", "1.0")

--- a/smithy-model/src/test/resources/software/amazon/smithy/model/loader/assembler-no-manifest-jar/META-INF/MANIFEST.MF
+++ b/smithy-model/src/test/resources/software/amazon/smithy/model/loader/assembler-no-manifest-jar/META-INF/MANIFEST.MF
@@ -1,0 +1,1 @@
+Manifest-Version: 1.0

